{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ecfe6fc7-be28-4e62-9319-6df7cedc50e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ans 01:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a63c96fc-f695-43d1-837d-385b0649b1c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# By inheriting from the Exception class or one of its subclasses, you inherit all the behaviors and features of those classes. This includes proper\n",
    "# error handling, traceback generation, and consistent behavior in Python's exception handling mechanisms.\n",
    "# Using the Exception class as the superclass when creating a custom exception in Python ensures that your custom exceptions play nicely with the rest\n",
    "# of the exception framework. It makes your code more organized and your custom exceptions easier to work with alongside built-in exceptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4081dc75-c3f2-4825-b10b-84d85a9f4ca1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a50e4ba3-9b26-4dca-9d43-9a296170e8bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ans 02:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ed8b395a-e857-44c5-9c0b-86f62a2175ab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "  <class 'BaseExceptionGroup'>\n",
      "    <class 'ExceptionGroup'>\n",
      "  <class 'Exception'>\n",
      "    <class 'ArithmeticError'>\n",
      "      <class 'FloatingPointError'>\n",
      "      <class 'OverflowError'>\n",
      "      <class 'ZeroDivisionError'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "      <class 'decimal.DecimalException'>\n",
      "        <class 'decimal.Clamped'>\n",
      "        <class 'decimal.Rounded'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Inexact'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Subnormal'>\n",
      "          <class 'decimal.Underflow'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'decimal.InvalidOperation'>\n",
      "          <class 'decimal.ConversionSyntax'>\n",
      "          <class 'decimal.DivisionImpossible'>\n",
      "          <class 'decimal.DivisionUndefined'>\n",
      "          <class 'decimal.InvalidContext'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'AttributeError'>\n",
      "      <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'EOFError'>\n",
      "      <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'ImportError'>\n",
      "      <class 'ModuleNotFoundError'>\n",
      "      <class 'zipimport.ZipImportError'>\n",
      "    <class 'LookupError'>\n",
      "      <class 'IndexError'>\n",
      "      <class 'KeyError'>\n",
      "        <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "        <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'NameError'>\n",
      "      <class 'UnboundLocalError'>\n",
      "    <class 'OSError'>\n",
      "      <class 'BlockingIOError'>\n",
      "      <class 'ChildProcessError'>\n",
      "      <class 'ConnectionError'>\n",
      "        <class 'BrokenPipeError'>\n",
      "        <class 'ConnectionAbortedError'>\n",
      "        <class 'ConnectionRefusedError'>\n",
      "        <class 'ConnectionResetError'>\n",
      "          <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'FileExistsError'>\n",
      "      <class 'FileNotFoundError'>\n",
      "      <class 'InterruptedError'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'IsADirectoryError'>\n",
      "      <class 'NotADirectoryError'>\n",
      "      <class 'PermissionError'>\n",
      "      <class 'ProcessLookupError'>\n",
      "      <class 'TimeoutError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'socket.herror'>\n",
      "      <class 'socket.gaierror'>\n",
      "      <class 'ssl.SSLError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'ssl.SSLZeroReturnError'>\n",
      "        <class 'ssl.SSLWantWriteError'>\n",
      "        <class 'ssl.SSLWantReadError'>\n",
      "        <class 'ssl.SSLSyscallError'>\n",
      "        <class 'ssl.SSLEOFError'>\n",
      "      <class 'shutil.Error'>\n",
      "        <class 'shutil.SameFileError'>\n",
      "      <class 'shutil.SpecialFileError'>\n",
      "      <class 'shutil.ExecError'>\n",
      "      <class 'shutil.ReadError'>\n",
      "      <class 'urllib.error.URLError'>\n",
      "        <class 'urllib.error.HTTPError'>\n",
      "        <class 'urllib.error.ContentTooShortError'>\n",
      "      <class 'gzip.BadGzipFile'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'RuntimeError'>\n",
      "      <class 'NotImplementedError'>\n",
      "        <class 'zmq.error.ZMQVersionError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "      <class 'RecursionError'>\n",
      "      <class '_frozen_importlib._DeadlockError'>\n",
      "      <class 'threading.BrokenBarrierError'>\n",
      "      <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "        <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "      <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "      <class 'asyncio.exceptions.BrokenBarrierError'>\n",
      "      <class 'pkg_resources.ExtractionError'>\n",
      "      <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'SyntaxError'>\n",
      "      <class 'IndentationError'>\n",
      "        <class 'TabError'>\n",
      "    <class 'SystemError'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'TypeError'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'ValueError'>\n",
      "      <class 'UnicodeError'>\n",
      "        <class 'UnicodeDecodeError'>\n",
      "        <class 'UnicodeEncodeError'>\n",
      "        <class 'UnicodeTranslateError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'json.decoder.JSONDecodeError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'binascii.Error'>\n",
      "      <class '_hashlib.UnsupportedDigestmodError'>\n",
      "      <class 'ipaddress.AddressValueError'>\n",
      "      <class 'ipaddress.NetmaskValueError'>\n",
      "      <class 'calendar.IllegalMonthError'>\n",
      "      <class 'calendar.IllegalWeekdayError'>\n",
      "      <class 'dateutil.parser._parser.ParserError'>\n",
      "      <class 'pygments.util.ClassNotFound'>\n",
      "      <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "      <class 'email.errors.MessageDefect'>\n",
      "        <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "        <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "        <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "        <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "        <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "        <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "        <class 'email.errors.UndecodableBytesDefect'>\n",
      "        <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "        <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "        <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "        <class 'email.errors.HeaderDefect'>\n",
      "          <class 'email.errors.InvalidHeaderDefect'>\n",
      "          <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "          <class 'email.errors.NonPrintableDefect'>\n",
      "          <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "          <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "          <class 'email.errors.InvalidDateDefect'>\n",
      "      <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "      <class 'plistlib.InvalidFileException'>\n",
      "      <class 'pkg_resources._vendor.more_itertools.recipes.UnequalIterablesError'>\n",
      "      <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "      <class 'pkg_resources._vendor.packaging._elffile.ELFInvalid'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "      <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "        <class 'pkg_resources.RequirementParseError'>\n",
      "      <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'Warning'>\n",
      "      <class 'BytesWarning'>\n",
      "      <class 'DeprecationWarning'>\n",
      "        <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "      <class 'EncodingWarning'>\n",
      "      <class 'FutureWarning'>\n",
      "        <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "      <class 'ImportWarning'>\n",
      "      <class 'PendingDeprecationWarning'>\n",
      "      <class 'ResourceWarning'>\n",
      "      <class 'RuntimeWarning'>\n",
      "        <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "        <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "        <class 'pkg_resources.PEP440Warning'>\n",
      "      <class 'SyntaxWarning'>\n",
      "      <class 'UnicodeWarning'>\n",
      "      <class 'UserWarning'>\n",
      "        <class 'getpass.GetPassWarning'>\n",
      "        <class 'IPython.core.formatters.FormatterWarning'>\n",
      "      <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "      <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'ExceptionGroup'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "      <class 'subprocess.CalledProcessError'>\n",
      "      <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "      <class 'concurrent.futures._base.CancelledError'>\n",
      "      <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class 'zlib.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "      <class 'zmq.error.ZMQError'>\n",
      "        <class 'zmq.error.ContextTerminated'>\n",
      "        <class 'zmq.error.Again'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'zmq.error.ZMQBindError'>\n",
      "      <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "      <class '_pickle.PicklingError'>\n",
      "      <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class '_ctypes.COMError'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "      <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "        <class 'traitlets.config.loader.ArgumentError'>\n",
      "      <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "      <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'configparser.Error'>\n",
      "      <class 'configparser.NoSectionError'>\n",
      "      <class 'configparser.DuplicateSectionError'>\n",
      "      <class 'configparser.DuplicateOptionError'>\n",
      "      <class 'configparser.NoOptionError'>\n",
      "      <class 'configparser.InterpolationError'>\n",
      "        <class 'configparser.InterpolationMissingOptionError'>\n",
      "        <class 'configparser.InterpolationSyntaxError'>\n",
      "        <class 'configparser.InterpolationDepthError'>\n",
      "      <class 'configparser.ParsingError'>\n",
      "        <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "      <class 'IPython.core.error.TryNext'>\n",
      "      <class 'IPython.core.error.UsageError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "      <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "      <class 'sqlite3.InterfaceError'>\n",
      "      <class 'sqlite3.DatabaseError'>\n",
      "        <class 'sqlite3.InternalError'>\n",
      "        <class 'sqlite3.OperationalError'>\n",
      "        <class 'sqlite3.ProgrammingError'>\n",
      "        <class 'sqlite3.IntegrityError'>\n",
      "        <class 'sqlite3.DataError'>\n",
      "        <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "      <class 'xml.dom.IndexSizeErr'>\n",
      "      <class 'xml.dom.DomstringSizeErr'>\n",
      "      <class 'xml.dom.HierarchyRequestErr'>\n",
      "      <class 'xml.dom.WrongDocumentErr'>\n",
      "      <class 'xml.dom.InvalidCharacterErr'>\n",
      "      <class 'xml.dom.NoDataAllowedErr'>\n",
      "      <class 'xml.dom.NoModificationAllowedErr'>\n",
      "      <class 'xml.dom.NotFoundErr'>\n",
      "      <class 'xml.dom.NotSupportedErr'>\n",
      "      <class 'xml.dom.InuseAttributeErr'>\n",
      "      <class 'xml.dom.InvalidStateErr'>\n",
      "      <class 'xml.dom.SyntaxErr'>\n",
      "      <class 'xml.dom.InvalidModificationErr'>\n",
      "      <class 'xml.dom.NamespaceErr'>\n",
      "      <class 'xml.dom.InvalidAccessErr'>\n",
      "      <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'IPython.core.guarded_eval.GuardRejection'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "      <class 'jedi.api.exceptions.InternalError'>\n",
      "      <class 'jedi.api.exceptions.WrongVersion'>\n",
      "      <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "      <class 'email.errors.MessageParseError'>\n",
      "        <class 'email.errors.HeaderParseError'>\n",
      "        <class 'email.errors.BoundaryError'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "      <class 'email.errors.CharsetError'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "      <class 'http.client.NotConnected'>\n",
      "      <class 'http.client.InvalidURL'>\n",
      "      <class 'http.client.UnknownProtocol'>\n",
      "      <class 'http.client.UnknownTransferEncoding'>\n",
      "      <class 'http.client.UnimplementedFileMode'>\n",
      "      <class 'http.client.IncompleteRead'>\n",
      "      <class 'http.client.ImproperConnectionState'>\n",
      "        <class 'http.client.CannotSendRequest'>\n",
      "        <class 'http.client.CannotSendHeader'>\n",
      "        <class 'http.client.ResponseNotReady'>\n",
      "      <class 'http.client.BadStatusLine'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "      <class 'psutil.NoSuchProcess'>\n",
      "        <class 'psutil.ZombieProcess'>\n",
      "      <class 'psutil.AccessDenied'>\n",
      "      <class 'psutil.TimeoutExpired'>\n",
      "    <class '_psutil_windows.Error'>\n",
      "    <class '_psutil_windows.TimeoutExpired'>\n",
      "    <class '_psutil_windows.TimeoutAbandoned'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "      <class 'xmlrpc.client.ProtocolError'>\n",
      "      <class 'xmlrpc.client.ResponseError'>\n",
      "      <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pkg_resources.extern.packaging._tokenizer.ParserSyntaxError'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "      <class 'pkg_resources.VersionConflict'>\n",
      "        <class 'pkg_resources.ContextualVersionConflict'>\n",
      "      <class 'pkg_resources.DistributionNotFound'>\n",
      "      <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "    <class 'pywintypes.error'>\n",
      "    <class 'pywintypes.com_error'>\n",
      "  <class 'GeneratorExit'>\n",
      "  <class 'KeyboardInterrupt'>\n",
      "  <class 'SystemExit'>\n",
      "  <class 'asyncio.exceptions.CancelledError'>\n",
      "  <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(\" \" * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 2)\n",
    "\n",
    "print_exception_hierarchy(BaseException)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f60b2ebc-793a-4cd6-afcd-4a8467b05e26",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a33748a2-9850-44e4-937d-672002ca7955",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ans 03:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "75271035-b110-4b6c-a6bb-dd8f93d48358",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The ArithmeticError class in Python is a base class for exceptions that are related to arithmetic operations. It serves as a superclass for\n",
    "# several more specific arithmetic-related exception classes. Some of the common exceptions that are defined as subclasses of ArithmeticError\n",
    "# include:\n",
    "\n",
    "# 1. ZeroDivisionError: Raised when division or modulo operation is performed with a divisor of zero.\n",
    "\n",
    "# 2. OverflowError: Raised when an arithmetic operation exceeds the limits of the data type.\n",
    "\n",
    "# 3. FloatingPointError: Raised when a floating-point arithmetic operation fails to produce a valid result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8071cd44-e5c7-48d0-907f-0745e3ee3750",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac48b615-afe8-4ea3-b5f0-1ddf6ed66fe6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ans 04:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ff04423e-2073-4ccf-bdfd-99a2c22fa214",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The LookupError class in Python is a base class for exceptions that are related to lookup operations, specifically when trying to access an element\n",
    "# in a collection or container and the element is not found or the index is out of range. It serves as a superclass for more specific lookup-related\n",
    "# exception classes.\n",
    "# Two common exceptions that are defined as subclasses of LookupError are KeyError and IndexError.\n",
    "\n",
    "# 1. KeyError: This exception is raised when trying to access a dictionary key that doesn't exist.\n",
    "\n",
    "# 2. IndexError: This exception is raised when trying to access a sequence (like a list or a string) using an invalid index, which is either negative or\n",
    "# greater than the length of the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f4d12907-041c-4e75-bda9-34d1f1266ebb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value: 2\n",
      "Caught KeyError: 'cherry'\n",
      "Value: 20\n",
      "Caught IndexError: list index out of range\n"
     ]
    }
   ],
   "source": [
    "def handle_key_error(dictionary, key):\n",
    "    try:\n",
    "        value = dictionary[key]\n",
    "        print(\"Value:\", value)\n",
    "    except KeyError as e:\n",
    "        print(\"Caught KeyError:\", e)\n",
    "\n",
    "def handle_index_error(sequence, index):\n",
    "    try:\n",
    "        value = sequence[index]\n",
    "        print(\"Value:\", value)\n",
    "    except IndexError as e:\n",
    "        print(\"Caught IndexError:\", e)\n",
    "\n",
    "# Test cases\n",
    "my_dict = {\"apple\": 2, \"banana\": 3}\n",
    "my_list = [10, 20, 30]\n",
    "\n",
    "handle_key_error(my_dict, \"apple\")\n",
    "handle_key_error(my_dict, \"cherry\")\n",
    "\n",
    "handle_index_error(my_list, 1)\n",
    "handle_index_error(my_list, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1e349b0b-a529-45b9-aa9d-c1401e1b3792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#####################################################################\n",
    "# Ans 05:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "60307e32-f4e1-4dc7-bd51-a44233aeb7c0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ImportError and ModuleNotFoundError are both exceptions in Python that occur when you have issues importing modules, but they have some differences\n",
    "# in terms of their use cases and behavior.\n",
    "\n",
    "# ImportError is a base class for exceptions that occur when there is a problem importing a module or a symbol (like a function or a class) from a module.\n",
    "# It can occur for various reasons, such as when the module or symbol doesn't exist, there's an issue with the module's code, or there's an issue with the\n",
    "# import statement itself. It can also be used as a base class to create more specific import-related exceptions.\n",
    "\n",
    "# ModuleNotFoundError is a subclass of ImportError.\n",
    "# It specifically indicates that the module being imported could not be found. This exception was introduced in Python 3.6 to provide more precise feedback\n",
    "# when a module cannot be located. In Python 3.6 and later versions, if you try to import a module that doesn't exist, you will get a ModuleNotFoundError."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6291045d-3c66-4180-9023-1626578fbbc4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#####################################################################\n",
    "# Ans 06:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "24c7bb9f-158c-4258-8ae1-b26fd54b5398",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 1. Be Specific in Exception Handling:\n",
    "# Catch only the exceptions you expect and can handle. Avoid using broad catch-all blocks like except Exception: as it might hide unexpected issues.\n",
    "# Use multiple except blocks for different types of exceptions you want to handle separately.\n",
    "\n",
    "# 2. Use Context Managers (with Statements):\n",
    "# Utilize the with statement and context managers (__enter__ and __exit__ methods) to ensure proper resource management, like closing files or network\n",
    "# connections automatically.\n",
    "\n",
    "# 3. Keep Exception Handling Local:\n",
    "# Place your try and except blocks as close as possible to where the exception might occur. This makes your code more readable and helps pinpoint the\n",
    "# source of errors.\n",
    "\n",
    "# 4. Don't Suppress Exceptions Unnecessarily:\n",
    "# Avoid catching exceptions and not doing anything meaningful with them. If you catch an exception, either handle it properly or let it propagate up\n",
    "# the call stack.\n",
    "\n",
    "# 5. Use finally for Cleanup:\n",
    "# When you need to perform cleanup tasks regardless of whether an exception occurred, use a finally block. It guarantees that the cleanup code will be\n",
    "# executed.\n",
    "\n",
    "# 6. Log Exceptions:\n",
    "# Log exceptions using the logging module to gather information about errors during runtime. This aids in debugging and understanding the flow of your\n",
    "# program.\n",
    "\n",
    "# 7. Handle Exceptions in Layers:\n",
    "# Handle exceptions at different levels of your application. For example, you might catch higher-level exceptions in a GUI application's UI layer and\n",
    "# more specific exceptions in lower-level components.\n",
    "\n",
    "# 8. Reraise Exceptions Judiciously:\n",
    "# If you catch an exception and cannot handle it properly, consider re-raising it using raise without any arguments. This preserves the original exception's\n",
    "# traceback for better debugging.\n",
    "\n",
    "# 9. Create Custom Exception Classes:\n",
    "# Create custom exception classes to represent specific errors in your application. This makes your code more readable and helps you distinguish different\n",
    "# error scenarios.\n",
    "\n",
    "# 10. Use the else Clause Sparingly:\n",
    "# The else clause in a try block is executed if no exceptions are raised. Use it for code that should only run if the try block succeeds without exceptions.\n",
    "\n",
    "# 11. Avoid Bare except Statements:\n",
    "# Avoid using except: without specifying the exception type. It makes your code less predictable and can lead to unintended behavior.\n",
    "\n",
    "# 12. Handle Expected Exceptions Gracefully:\n",
    "# If you expect a certain type of exception to be raised in a specific situation (e.g., file not found), handle it gracefully and provide informative error\n",
    "# messages to users.\n",
    "\n",
    "# 13. Document Exception Handling:\n",
    "# Document the expected exceptions, their meanings, and how they are handled in your code. This helps other developers understand your code and contribute\n",
    "# more effectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37683cf9-a9a8-404a-8b10-fcaf49db881a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################################################################"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
